src/connectors/queue/payTo/payToByBlockchain.ts

```diff
@@ -0,0 +1,364 @@
+    this.q.process(
+      QUEUE_JOB.payTo,
+      QUEUE_CONCURRENCY.payToByBlockchain,
+      this.handlePayTo
+    )
+  }
+
+  private updateTxAndBlockchainTxState = async (
+    {
+      txId,
+      txState,
+      txRemark,
+    }: {
+      txId: string
+      txState: TRANSACTION_STATE
+      txRemark?: TRANSACTION_REMARK
+    },
+    {
+      blockchainTxId,
+      blockchainTxState,
+    }: {
+      blockchainTxId: string
+      blockchainTxState: BLOCKCHAIN_TRANSACTION_STATE
+    }
+  ) => {
+    const trx = await this.knex.transaction()
+    try {
+      await this.paymentService.markTransactionStateAs(
+        {
+          id: txId,
+          state: txState,
+          remark: txRemark,
+        },
+        trx
+      )
+      await this.paymentService.markBlockchainTransactionStateAs(
+        {
+          id: blockchainTxId,
+          state: blockchainTxState,
+        },
+        trx
+      )
+      await trx.commit()
+    } catch (error) {
+      await trx.rollback()
+      throw error
+    }
+  }
+
+  private failBothTxAndBlockchainTx = async (
+    txId: string,
+    blockchainTxId: string
+  ) => {
+    await this.updateTxAndBlockchainTxState(
+      { txId, txState: TRANSACTION_STATE.failed },
+      {
+        blockchainTxId,
+        blockchainTxState: BLOCKCHAIN_TRANSACTION_STATE.reverted,
+      }
+    )
+  }
+
+  private timeoutBlockchainTx = async (blockchainTxId: string) => {
+    await this.paymentService.markBlockchainTransactionStateAs({
+      id: blockchainTxId,
+      state: BLOCKCHAIN_TRANSACTION_STATE.timeout,
+    })
+  }
+
+  private validateTxLogs = async (
+    logs: Log[],
+    {
+      curatorAddress,
+      creatorAddress,
+      cid,
+      tokenAddress,
+      amount,
+      decimals,
+    }: {
+      curatorAddress?: string
+      creatorAddress?: string
+      cid: string
+      tokenAddress: string
+      amount: string
+      decimals: number
+    }
+  ) => {
+    const abi = [
+      'event Curation(address indexed curator, address indexed creator, address indexed token, string uri, uint256 amount)',
+    ]
+    const topic =
+      '0xc2e41b3d49bbccbac6ceb142bad6119608adf4f1ee1ca5cc6fc332e0ca2fc602'
+    if (logs.length === 0) {
+      return false
+    } else {
+      for (const log of logs) {
+        if (
+          log.address.toLowerCase() ===
+            environment.curationContractAddress.toLowerCase() &&
+          log.topics[0] === topic
+        ) {
+          const iface = new ethers.utils.Interface(abi)
+          const event = iface.parseLog(log)
+          if (
+            event.args.curator!.toLowerCase() ===
+              curatorAddress!.toLowerCase() &&
+            event.args.creator!.toLowerCase() ===
+              creatorAddress!.toLowerCase() &&
+            event.args.token!.toLowerCase() === tokenAddress.toLowerCase() &&
+            event.args.amount!.toString() ===
+              toTokenBaseUnit(amount, decimals) &&
+            (event.args.uri!.includes(cid) ||
```

Comment:
`uri` will be `ipfs://{CID}`, so here we have to strip out the prefix `ipfs://`?

Reply:
if some third parties call contract directly（and call matters graphQL api at the same time. Oh my god ）， we can't make sure they follow the same rule (it could be a gateway url).  

here may be we should validate the uri schema first

Reply:
we can check if `uri` has the `ipfs://`. If true, check if it's a valid CID (i think this check be omit) and match with the db record.

Reply:
```ts
if ((/^ipfs:\/\//.test(uri))) {
   cid = uri.replace('ipfs://','')
}
```